adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,253 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'opentelemetry/trace/status'
+
+require_relative '../util/queue_time'
+
+module OpenTelemetry
+  module Adapters
+    module Rack
+      module Middlewares
+        # TracerMiddleware propagates context and instruments Rack requests
+        # by way of its middleware system
+        #
+        # Notable implementation differences from dd-trace-rb:
+        # * missing: 'span.resource', which is set to span.name
+        # * missing: config[:distributed_tracing]
+        # * missing: span.set_error() -- spec level change
+        class TracerMiddleware # rubocop:disable Metrics/ClassLength
+          class << self
+            def allowed_rack_request_headers
+              @allowed_rack_request_headers ||= Array(config[:allowed_request_headers]).each_with_object({}) do |header, memo|
+                memo["HTTP_#{header.to_s.upcase.gsub(/[-\s]/, '_')}"] = build_attribute_name('http.request.headers.', header)
+              end
+            end
+
+            def allowed_response_headers
+              @allowed_response_headers ||= Array(config[:allowed_response_headers]).each_with_object({}) do |header, memo|
+                memo[header] = build_attribute_name('http.response.headers.', header)
+                memo[header.to_s.upcase] = build_attribute_name('http.response.headers.', header)
+              end
+            end
+
+            def build_attribute_name(prefix, suffix)
+              prefix + suffix.to_s.downcase.gsub(/[-\s]/, '_')
+            end
+
+            def config
+              Rack::Adapter.instance.config
+            end
+
+            private
+
+            def clear_cached_config
+              @allowed_rack_request_headers = nil
+              @allowed_response_headers = nil
+            end
+          end
+
+          EMPTY_HASH = {}.freeze
+
+          def initialize(app)
+            @app = app
+          end
+
+          def call(env)
+            extracted_context = OpenTelemetry.propagation.extract(env)
+            frontend_context = create_frontend_span(env, extracted_context)
+            request_context = create_request_span(env, frontend_context || extracted_context)
+            request_span = request_context[current_span_key]
+
+            # restore extracted context in this process:
+            OpenTelemetry::Context.with_current(request_context) do
+              begin
+                @app.call(env).tap do |status, headers, response|
+                  set_attributes_after_request(request_span, status, headers, response)
+                end
+              rescue StandardError => e
+                record_and_reraise_error(e, request_span: request_span)
```

Comment:
Now that `in_span` records and reraises errors, you could consider rewriting call as follows:

```ruby
def call(env)
            original_env = env.dup
            extracted_context = OpenTelemetry.propagation.extract(env)
            frontend_context = create_frontend_span(env, extracted_context)

            OpenTelemetry::Context.with_current(frontend_context || extracted_context) do
              request_span_name = create_request_span_name(env['REQUEST_URI'] || original_env['PATH_INFO'])
              tracer.in_span(request_span_name,
                             attributes: request_span_attributes(env: env),
                             kind: :server) do |request_span|
                @app.call(env).tap do |status, headers, response|
                  set_attributes_after_request(request_span, status, headers, response)
                end
              end
            end
          ensure
            finish_span(frontend_context)
          end
```

Reply:
I've rewritten, as suggested :+1: